core/src/main/java/com/linecorp/armeria/client/ClientRequestContext.java

```diff
@@ -138,9 +142,29 @@ static ClientRequestContext of(RpcRequest request, URI uri) {
     @Override
     ClientRequestContext newDerivedContext(Request request);
 
+    /**
+     * Creates a new derived {@link RequestContext} with the specified {@link Request} and {@link Endpoint}
+     * which the {@link RequestLog} is different from the deriving context.
+     * Note that the references of {@link Attribute}s in the {@link #attrs()} are copied as well.
+     */
+    ClientRequestContext newDerivedContext(Request request, Endpoint endpoint);
+
+    /**
+     * Returns the {@link EndpointSelector} used for the current {@link Request}.
+     *
+     * @return the {@link EndpointSelector} if a user specified a group {@link Endpoint}
+     *         {@code null} if a user specified a host {@link Endpoint}.
+     */
+    @Nullable
+    EndpointSelector endpointSelector();
+
     /**
      * Returns the remote {@link Endpoint} of the current {@link Request}.
+     *
+     * @return the remote {@link Endpoint}. {@code null} if the {@link Request} has failed
+     *         before its remote {@link Endpoint} is determined.
```

Comment:
Is it better using `before` or `because its remote endpoint couldn't be determined`?

Reply:
Changed to `{@code null} if the {@link Request} has failed because its remote {@link Endpoint} couldn't be determined.`